# syntax-transform/3.1/0029-retry-outside-rescue.patch
#
# This file is part of mkxp.
#
# Copyright (C) 2026 The mkxp-z authors
#
# mkxp is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 2 of the License, or
# (at your option) any later version.
#
# mkxp is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with mkxp.  If not, see <http://www.gnu.org/licenses/>.

# In Ruby 1.8, the `retry` keyword can be used inside a loop or block to
# restart it from the beginning, behaving identically to `redo`. For example,
# this code snippet in Ruby 1.8 prints `0` indefinitely:
#
#   for i in 0..10
#     p i
#     retry if i == 1
#   end
#
# In Ruby 1.9 and later, `retry` outside of a `rescue` clause is a compile
# error ("Invalid retry"). This patch makes `retry` outside of `rescue` behave
# as `redo` when syntax transform is active and targeting Ruby <= 1.8.

--- a/compile.c
+++ b/compile.c
@@ -7532,6 +7532,11 @@ compile_redo(rb_iseq_t *iseq, LINK_ANCHOR *const ret, const NODE *const node, in
 static int
 compile_retry(rb_iseq_t *iseq, LINK_ANCHOR *const ret, const NODE *const node, int popped)
 {
+    /* If syntax transform is active and targeting Ruby <= 1.8, treat
+     * `retry` outside of rescue as `redo` to match Ruby 1.8's behavior. */
+    if (iseq->body->type != ISEQ_TYPE_RESCUE && mkxp_iseq_is_syntax_transform_active(iseq, 1, 8, -1)) {
+        return compile_redo(iseq, ret, node, popped);
+    }
     const NODE *line_node = node;
 
     if (iseq->body->type == ISEQ_TYPE_RESCUE) {
